libs/ml/include/ml/model/model.hpp

```diff
@@ -272,4 +284,214 @@ void Model<TensorType>::PrintStats(SizeType epoch, DataType loss, DataType test_
 
 }  // namespace model
 }  // namespace ml
+
+namespace serializers {
+/**
+ * serializer for Model
+ * @tparam TensorType
+ */
+template <typename TensorType, typename D>
+struct MapSerializer<ml::model::Model<TensorType>, D>
+{
+  using Type       = ml::model::Model<TensorType>;
+  using DriverType = D;
+
+  // protected member variables
+  static uint8_t const GRAPH           = 1;
+  static uint8_t const MODEL_CONFIG    = 2;
+  static uint8_t const DATALOADER_PTR  = 3;
+  static uint8_t const DATALOADER_TYPE = 4;
+  static uint8_t const OPTIMISER_PTR   = 5;
+  static uint8_t const OPTIMISER_TYPE  = 6;
+
+  static uint8_t const INPUT_NODE_NAME  = 7;
+  static uint8_t const LABEL_NODE_NAME  = 8;
+  static uint8_t const OUTPUT_NODE_NAME = 9;
+  static uint8_t const ERROR_NODE_NAME  = 10;
+
+  static uint8_t const LOSS_SET_FLAG      = 11;
+  static uint8_t const OPTIMISER_SET_FLAG = 12;
+  static uint8_t const COMPILED_FLAG      = 13;
+
+  template <typename MapType>
+  static void SerializeDataLoader(MapType map, Type const &sp)
+  {
+    map.Append(DATALOADER_TYPE, static_cast<uint8_t>(sp.dataloader_ptr_->LoaderCode()));
+
+    switch (sp.dataloader_ptr_->LoaderCode())
```

Comment:
is this switch case code really necessary?
I think you might be able to just have the main Serialize() function with map.Append(). Specifically, because some cases may not be implemented yet you may want to do this:
try
{
  map.Append(sp.dataloader_)
}
catch (Exception& e)
{
   cout << e.what() << endl;
}

If it doesn't compile because it requires a Serialise function to exist for each dataloader then just create them with serialise and deserialise functions that simply throw ml::exceptions::NotYetImplementedError()